### DISCUSSION TITLE
outputFileTracingIncludes in app directory routes

### DISCUSSION THREAD

**Discussion by theAJFM** [vercel/next.js#55228] [App Router]
Hi,

I've yet to figure out a way to force files to be included in the output file tracing for a given path. In the page directory, I was able to do something like this:

```js
outputFileTracingIncludes: {
  '/api/post': ['.next/server/chunks/*.wasm'],
},
```
 
In the app directory, doing something like this in the new API route handler doesn't work:

```javascript
outputFileTracingIncludes: {
  '/api/post/route': ['.next/server/chunks/*.wasm'],
},
```

I used `/api/post/route` since it's the name of the file that was traced, and after looking into its `nft.json` file, the wasm files never appeared. 

From looking at the Next.js source [code](https://github.com/vercel/next.js/blob/df76f6633fe192e04b88ace62a00ccc9bb833d86/packages/next/src/build/index.ts#L1969), I get a sense that this is a feature that's only available in the pages directory:

```js
const traceFile = path.join(
  distDir,
  'server/pages', // This is the parent folder for all of the page route pages and api routes alike in the build output. App dir page files are placed elsewhere.
  `${page}.js.nft.json`
)
```

Is it true that this is only available in pages directory? Any plan to enable this in the app directory as well?

**Comment by sgoodrow-zymergen** [comment]
+1

**Comment by vinnymac** [comment]
It has been a year, and I was able to get this working after some trial/error.
I am using NextJS 14.2.15 and Turborepo 2.0.5, and my config looks like

```ts
const config = {
  experimental: {
    outputFileTracingIncludes: {
      '/api/nft': ['./src/asset/path/*'],
    },
  },
};
```

I tried `/api/nft/*` and `/api/nft/route` as well, and neither of those worked. I confirmed each of my attempts by going to `.next/server/app/nft/route.js.nft.json` and verifying whether or not they included `src/asset/path` files. Only `/api/nft` produced the appropriate traces.

https://nextjs.org/docs/app/api-reference/next-config-js/output#how-it-works

Note: The [vercel.json](https://vercel.com/docs/projects/project-configuration#functions) functions `includeFiles` docs did not work for me, whether or not I was using vercel to host my NextJS app.

Further, the project I was working on had another issue where a `fileURLToPath(import.meta.url)` was used, switching to [process.cwd()](https://vercel.com/guides/how-can-i-use-files-in-serverless-functions#examples-of-reading-files) fixed it, despite ESM working just fine locally.

**Reply by wscourge** [reply]
Thank you, this helped me fix all my Static and SSG pages relying on `process.cwd()` on Vercel. Everything worked as expected when building it locally.

One thing that wasn't obvious is that `experimental.outputFileTracingIncludes` only works together with `output: "standalone"` - I had to update. So the final _next.config.js_ was:

```
const config = {
  output: "standalone",
  experimental: {
    outputFileTracingIncludes: {
      '/api/nft': ['./src/asset/path/*'],
    },
  },
};
```

**Reply by vinnymac** [reply]
> One thing that wasn't obvious is that experimental.outputFileTracingIncludes only works together with output: "standalone"

That is not the case, I have never used output standalone while using outputFileTracingIncludes. It might depend on whether you’re hosted on Vercel? Not sure of the differences between our setups, but the code above was used in a barebones NextJS app deployed to Vercel.
